### PR TITLE
packet/bgp: Support Revised Error Handling (RFC7606)

### LINKED ISSUE #1147: Malformed large BGP communities attributes handling
I would like to ask your support to analyze a behaviour that seems unexpected to me. It seems GoBGP does not handle malformed large BGP communities attributes in the way I understand the draft wants to.

Draft:

> A BGP UPDATE message with a malformed Large Communities attribute SHALL be handled using the approach of "treat-as-withdraw" as described in section 2 [RFC7606].

RFC7606:

> Treat-as-withdraw: In this approach, the UPDATE message containing the path attribute in question MUST be treated as though all contained routes had been withdrawn just as if they had been listed in the WITHDRAWN ROUTES field (or in the MP_UNREACH_NLRI attribute if appropriate) of the UPDATE message, thus causing them to be removed from the Adj-RIB-In according to the procedures of [RFC4271].

I used an ExaBGP instance to send a malformed attribute; this is the output of tcpdump on the GoBGP host:

```
    192.0.2.2.40208 > 192.0.2.3.179: Flags [P.], cksum 0x8474 (incorrect -> 0x2f47), seq 350:422, ack 65, win 457, options [nop,nop,TS val 5124424 ecr 5123924], length 72: BGP, length: 72
        Update Message (2), length: 72
          Origin (1), length: 1, Flags [T]: IGP
          AS Path (2), length: 6, Flags [T]: 65536
          Next Hop (3), length: 4, Flags [T]: 192.0.2.2
          Unknown Attribute (30), length: 21, Flags [OTP]:
            no Attribute 30 decoder
            0x0000:  0000 0001 0000 0002 0000 0003 0000 0004
            0x0010:  0000 0005 ff
          Updated routes:
            203.0.113.12/32
    192.0.2.3.179 > 192.0.2.2.40208: Flags [P.], cksum 0x8441 (incorrect -> 0xf9e0), seq 65:86, ack 422, win 470, options [nop,nop,TS val 5124425 ecr 5124424], length 21: BGP, length: 21
        Notification Message (3), length: 21, UPDATE Message Error (3), subcode Attribute Length Error (5)
    192.0.2.3.179 > 192.0.2.2.40208: Flags [F.], cksum 0x842c (incorrect -> 0x0200), seq 86, ack 422, win 470, options [nop,nop,TS val 5124425 ecr 5124424], l
```

This is GoBGP's log:

```
WARN[3024] malformed BGP message  Key=192.0.2.2 State=BGP_FSM_ESTABLISHED Topic=Peer error=large communities length isn't correct
WARN[3024] sent notification      Data=&{Header:{Marker:[] Len:21 Type:3} Body:0xc4203f5040} Key=192.0.2.2 State=BGP_FSM_ESTABLISHED Topic=Peer
INFO[3024] Peer Down              Key=192.0.2.2 Reason=notification-sent code 3(update) subcode 5(attribute length error) State=BGP_FSM_ESTABLISHED Topic=Peer
```

It seems that GoBGP, instead of treating the message as a withdraw, drops the BGP session with the sending neighbor.

Thanks.

GoBGP config:

```
[global.config]
        as = 65537
        router-id = "192.0.2.3"

[[neighbors]]
        [neighbors.config]
                neighbor-address = "192.0.2.2"
                peer-as = 65536
```

### PR DESCRIPTION
**Pull request by fujimoto-s**
This patch enables GoBGP to keep the session established even if the received Update message contains some errors, and to handle these errors according to what defined in RFC7606.

This feature is enabled when 'treat-as-withdraw' in 'neighbors.error-handling.config' is specified to true in the GoBGP config file.

This PR fixes #1147.

### CODE REVIEW

File: packet/bgp/bgp.go [hunk 1]
```diff
@@ -7870,20 +7871,46 @@ func (msg *BGPUpdate) DecodeFromBytes(data []byte, options ...*MarshallingOption
...
 		}
+
 		err = p.DecodeFromBytes(data, options...)
 		if err != nil {
-			return err
+			e = err.(*MessageError)
+			if e.(*MessageError).SubTypeCode == BGP_ERROR_SUB_ATTRIBUTE_FLAGS_ERROR {
+				e.(*MessageError).ErrorHandling = ERROR_HANDLING_TREAT_AS_WITHDRAW
+			} else {
+				e.(*MessageError).ErrorHandling = PathAttributeToErrorHandlingMap[p.GetType()]
```
Comment (fujita) [inline packet/bgp/bgp.go]: Out of index with PathAttributeUnknown?

Comment (fujimoto-s) [reply packet/bgp/bgp.go]: Oh, it will happen.
I'll fix it, thanks!

File: packet/bgp/bgp.go [hunk 2]
```diff
@@ -7982,6 +7982,33 @@ func (msg *BGPUpdate) IsEndOfRib() (bool, RouteFamily) {
...
+	var unreach []AddrPrefixInterface
+
+	for _, p := range msg.PathAttributes {
+		switch nlri := p.(type) {
+		case *PathAttributeMpReachNLRI:
+			unreach = append(unreach, nlri.Value...)
+		case *PathAttributeMpUnreachNLRI:
+			unreach = append(unreach, nlri.Value...)
+		default:
+			withdraw.PathAttributes = append(withdraw.PathAttributes, p)
```
Comment (fujita) [inline packet/bgp/bgp.go]: Why necessary? Withdraw message doesn't need any attributes.

**Comment by fujita** [comment]
Can you add some tests for this?

File: packet/bgp/bgp.go (hunk 2 above)
Comment (fujimoto-s) [reply packet/bgp/bgp.go]: It is not necessary, sorry.
I will remove it.

**Comment by fujimoto-s** [comment]
That would be fine. I will consider adding both unit tests and scenario tests.

**Comment by fujimoto-s** [comment]
I've updated my PR with additional unit/scenario tests.

**Comment by fujita** [comment]
Thanks!

**Comment by fujita** [comment]
@fujimoto-s We can close this too?
https://github.com/osrg/gobgp/issues/1143#issuecomment-257372397

**Comment by fujimoto-s** [comment]
No, it isn't.

RFC7606 says that the duplicate attribute should be discarded.
Thus, "handling the duplicate Large Communities attribute in an UPDATE message" is in the scope of RFC7606.
On the other hand, "handling the duplicate Large Communities __in one Large Communities attribute__" is out of scope.
I think we should make new PR to fix #1143.

**Comment by fujita** [comment]
I see, can you make a new PR?

**Comment by fujimoto-s** [comment]
Sure, I'll work on it.

**Pull request closed without merging**